feat: Hybrid SDK options and React Native manual native initialization page (#3818)
getsentry/sentry-docs: src/platforms/react-native/manual-setup/native-init.mdx

---
title: Native Initialization
sidebar_order: 4
description: "Learn how to manually initialize the native SDKs."
---

> By default, the React Native SDK will initialize the native SDK underneath when the `init` method is called on the JS layer. This currently has the limitation of not capturing native crashes that occur before the `init` method is called on the JS layer. You could also look to initialize the native SDKs yourself to be able to provide custom options above what the React Native SDK currently provides.

Reviewer: It seems like initializing the SDK manually is what you're telling the user to do instead of letting the default initialization process happen, so I tried to make the contrast clearer. 

```suggestion
By default, the React Native SDK will initialize the native SDK underneath when the `init` method is called on the JavaScript layer. However, this method currently has the limitation of not capturing native crashes that occur before the `init` method is called on the JavaScript layer. To overcome this, you can initialize the native SDKs manually to provide custom options beyond what the React Native SDK currently provides.
```

Author: Hmm this one is an either or, so you maybe also choose to do this if you want to provide custom options.

> By default, the React Native SDK will initialize the native SDK underneath when the `init` method is called on the JS layer. This currently has the limitation of not capturing native crashes that occur before the `init` method is called on the JS layer. You could also look to initialize the native SDKs yourself to be able to provide custom options above what the React Native SDK currently provides.

Reviewer 2: hmm. I found this a little confusing, so perhaps:
```suggestion
By default, the React Native SDK initializes the native SDK underneath the `init` method called on the JS layer. As a result, the SDK has a current limitation of not capturing native crashes that occur prior to the `init` method being called on the JS layer. You can initialize the native SDKs yourself to overcome this limitation or if you want to provide custom options above what the React Native SDK currently provides.
```

> To do so, you would set [autoInitializeNativeSdk](/platforms/react-native/configuration/options/#autoInitializeNativeSdk) to `false` in the init options:

Reviewer: ```suggestion
To do this, set [autoInitializeNativeSdk](/platforms/react-native/configuration/options/#autoInitializeNativeSdk) to `false` in the init options:
```

```javascript
Sentry.init({
  dsn: "___PUBLIC_DSN___",
  autoInitializeNativeSdk: false,
});
```

This will prevent the React Native SDK from initializing the native SDK automatically.

> <Alert level="info" title="Note">

Reviewer 2: When using "Alert", the box is red and eye catching, so it shouldn't be labeled "Note". If it's just a Note, use the Note/Note construction. If you want to keep this as a red, eye-catching box, change the title to "Disabling the Native Layer" or something more useful that helps the reader understand why we're calling this out so prominently.

> Do not use `autoInitializeNativeSdk` to disable the native layer, but instead use [enableNative](/platforms/react-native/configuration/options/#enableNative)

Reviewer 2: ```suggestion
Do not use `autoInitializeNativeSdk` to disable the native layer, instead use [enableNative](/platforms/react-native/configuration/options/#enableNative)
```

> Do not use `autoInitializeNativeSdk` to disable the native layer, instead use [enableNative](/platforms/react-native/configuration/options/#enableNative)

Reviewer: ```suggestion
Do not use `autoInitializeNativeSdk` to disable the native layer, but instead use [enableNative](/platforms/react-native/configuration/options/#enableNative)
```

</Alert>

You will then have to initialize the native SDKs yourself following the guides for Android or iOS. You do not need to install any of the native SDKs by yourself as they are already packaged with the React Native SDK.

## Android Configuration

> To use auto-init, you would have to add the following to your `AndroidManifest.xml`.

Reviewer: ```suggestion
To use auto-init, add the following to your `AndroidManifest.xml`.
```

```xml
<meta-data android:name="io.sentry.auto-init" tools:replace="android:value" android:value="true" />
```

> then you can follow [the guide on initializing Android](/platforms/android/#configure).

Reviewer: ```suggestion
Then follow [the guide on initializing Android](/platforms/android/#configure).
```

<Note>

> Initializing via the above method uses a Content Provider; because the Javascript engine for React Native is also initialized via a Content Provider we cannot guarantee that Sentry would always initialize before the Javascript engine.

Reviewer: we avoid latin terms

```suggestion
Initializing with the above method uses a Content Provider. Because the Javascript engine for React Native is also initialized using a Content Provider, we cannot guarantee that Sentry will always initialize before the JavaScript engine.
```

</Note>

> If you chose to manually initialize the Android SDK you can [follow the guide](/platforms/android/configuration/manual-init/) without needing to add the tag as `io.sentry.auto-init` is set to `false` by the React Native SDK.

Reviewer: ```suggestion
If you chose to manually initialize the Android SDK, you can [follow the guide](/platforms/android/configuration/manual-init/) without needing to add the tag as `io.sentry.auto-init` is set to `false` by the React Native SDK.
```

> You will then have to initialize the native SDKs yourself following the guides for Android or iOS. You do not need to install any of the native SDKs by yourself as they are already packaged with the React Native SDK.
> 
> ## Android Configuration
> 
> To use auto-init, add the following to your `AndroidManifest.xml`.
> 
> ```xml
> <meta-data android:name="io.sentry.auto-init" tools:replace="android:value" android:value="true" />
> ```
> 
> Then follow [the guide on initializing Android](/platforms/android/#configure).
> 
> <Note>
> 
> Initializing with the above method uses a Content Provider. Because the Javascript engine for React Native is also initialized using a Content Provider, we cannot guarantee that Sentry will always initialize before the JavaScript engine.
> 
> </Note>
> 
> If you chose to manually initialize the Android SDK, you can [follow the guide](/platforms/android/configuration/manual-init/) without needing to add the tag as `io.sentry.auto-init` is set to `false` by the React Native SDK.
> 
> ## iOS Configuration
> 
> You can just follow the [configuration guide for iOS](/platforms/apple/guides/ios/#configure) to initialize the Sentry Cocoa SDK.

Reviewer 2: Since the iOs configuration is simply a link, maybe we can shorten this up a bit?
```suggestion
Next, initialize the native SDKs for Android, which is documented below, or for iOS by following the [configuration guide to initialize the Sentry Cocoa SDK](/platforms/apple/guides/ios/#configure). Note that you do not need to install the native SDKs as they are already packaged with the React Native SDK.

## Android Configuration

<Note>

Initializing with this method uses a Content Provider. Because the Javascript engine for React Native is also initialized using a Content Provider, we cannot guarantee that Sentry will always initialize before the JavaScript engine.

</Note>

To use auto-init, add the following to your `AndroidManifest.xml`:

```xml
<meta-data android:name="io.sentry.auto-init" tools:replace="android:value" android:value="true" />
```

Then follow [the guide on initializing Android](/platforms/android/#configure).

If you manually initialized the Android SDK, you can then [follow the guide](/platforms/android/configuration/manual-init/) without needing to add the tag since `io.sentry.auto-init` is set to `false` by the React Native SDK.

```
